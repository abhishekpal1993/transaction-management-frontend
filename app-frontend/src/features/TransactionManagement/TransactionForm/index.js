import { Button } from "../../../components/atoms";
import { Input } from "../../../components/molecules";
import { Form } from "../../../components/renderprops";

import { schema, initialValue } from "./static";

export const TransactionForm = ({ onSubmit, isLoading }) => {
  return (
    <form
      className="w-full border py-4 px-4 rounded"
      onSubmit={(e) => e.preventDefault()}
    >
      <Form
        validateOnBlur={true}
        initialValues={initialValue}
        validationSchema={schema}
        onSubmitHandler={onSubmit}
      >
        {({
          values,
          errors,
          touched,
          handleChange,
          handleSubmit,
          handleBlur,
        }) => (
          <>
            <Input
              type="text"
              labelClassName="mb-4"
              data-type="account-id"
              label="Account ID:"
              name="accountId"
              value={values.accountId ?? ""}
              onChange={handleChange}
              onBlur={handleBlur}
              isError={!!errors.accountId && !!touched.accountId}
              error={errors.accountId}
            />
            <Input
              type="number"
              labelClassName="mb-4"
              data-type="amount"
              label="Amount:"
              name="amount"
              value={values.amount ?? 0}
              onChange={handleChange}
              onBlur={handleBlur}
              isError={!!errors.amount && !!touched.amount}
              error={errors.amount}
            />
            <Button
              className="w-full"
              data-type="transaction-submit"
              onClick={handleSubmit}
              isLoading={isLoading}
              disabled={
                !values.amount ||
                !values.accountId ||
                errors.amount ||
                errors.accountId
              }
            >
              Submit
            </Button>
          </>
        )}
      </Form>
    </form>
  );
};
