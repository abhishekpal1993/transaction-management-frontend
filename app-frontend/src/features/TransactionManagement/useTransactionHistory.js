import { useEffect, useState, useMemo, useCallback } from "react";
import {
  fetchAllData,
  getAccountById,
  getTransactionById,
  createTransaction,
} from "../../services/accounting";
import { BACKEND_RESPONSE_KEYS } from "../../common/constants";

export const useTransactionHistory = (props) => {
  const [transactions, setTransactions] = useState(props.transactions);
  const [accounts, setAccounts] = useState(props.accounts);
  const [isLoading, setLoading] = useState(false);

  const refetchData = useCallback(async () => {
    try {
      const res = await fetchAllData();
      setTransactions(res.transactions);
      setAccounts(res.accounts);
    } catch (err) {
      console.error(err);
    }
  }, []);

  const transactionList = useMemo(
    () =>
      transactions.map((trnx) => ({
        ...trnx,
        balance: accounts[trnx[BACKEND_RESPONSE_KEYS.accountId]],
      })),
    [transactions, accounts]
  );

  const onSubmitHandler = useCallback(async (values) => {
    try {
      setLoading(true);
      const { data: createTrnxRes } = await createTransaction(values);
      const [transaction, account] = await Promise.all([
        getTransactionById(
          createTrnxRes[BACKEND_RESPONSE_KEYS.transactionId]
        ).then((res) => res.data),
        getAccountById(createTrnxRes[BACKEND_RESPONSE_KEYS.accountId]).then(
          (res) => res.data
        ),
      ]);

      setTransactions((st) => [transaction, ...st]);
      setAccounts((accts) => ({
        ...accts,
        [account[BACKEND_RESPONSE_KEYS.accountId]]: account.balance,
      }));
    } catch (err) {
      console.error(err);
    } finally {
      setLoading(false);
    }
  }, []);

  // If prefetching data failed then try to fetch data on browser
  useEffect(() => {
    (async () => {
      if (!props.isPrefetched) {
        await refetchData();
      }
    })();
  }, []);

  return { transactionList, onSubmitHandler, isLoading };
};
