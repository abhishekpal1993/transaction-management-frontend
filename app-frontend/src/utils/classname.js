export const append = (...classNames) =>
  classNames.filter((cls) => !!cls).join(" ");
