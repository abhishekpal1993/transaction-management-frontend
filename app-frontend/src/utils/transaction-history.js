export const getTransactionTextForDebit = (amount, accountId, currency = "$") =>
  `Transferred ${currency}${amount} from account ${accountId}`;

export const getTransactionTextForCredit = (
  amount,
  accountId,
  currency = "$"
) => `Transferred ${currency}${amount} to account ${accountId}`;

export const getAmountSign = (amount) => (Math.sign(amount) === -1 ? "-" : "");

export const getCurrentAccountBalanceText = (amount, currency = "$") =>
  `The current account balance ${getAmountSign(amount)}${currency}${Math.abs(
    amount
  )}`;
