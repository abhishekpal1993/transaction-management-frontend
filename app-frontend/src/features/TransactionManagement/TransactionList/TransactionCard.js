import { Card, CardText } from "../../../components/atoms";
import { append } from "../../../utils/classname";
import {
  getCurrentAccountBalanceText,
  getTransactionTextForCredit,
  getTransactionTextForDebit,
} from "../../../utils/transaction-history";

export const TransactionCard = ({
  className,
  accountId,
  amount,
  currentAccountBalance,
  showBalance = false,
  ...props
}) => {
  const transactionCardClassName = "flex flex-col items-start";
  const cardText =
    amount < 0
      ? getTransactionTextForDebit(amount, accountId)
      : getTransactionTextForCredit(amount, accountId);
  const currentBalanceText = getCurrentAccountBalanceText(
    currentAccountBalance
  );
  return (
    <Card
      data-type="transaction"
      data-account-id={accountId}
      data-amount={amount}
      data-balance={currentAccountBalance}
      className={append(transactionCardClassName, className)}
      {...props}
    >
      <CardText>{cardText}</CardText>
      {!!showBalance && (
        <CardText className="mt-4">{currentBalanceText}</CardText>
      )}
    </Card>
  );
};
