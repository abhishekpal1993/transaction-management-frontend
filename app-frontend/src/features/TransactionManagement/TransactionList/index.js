import { TransactionCard } from "./TransactionCard";
import { BACKEND_RESPONSE_KEYS } from "../../../common/constants";

export const TransactionList = ({ data }) => {
  if (!Array.isArray(data) || data.length < 1) {
    return <div>No Transactions!</div>;
  }

  const [firstItem, ...rest] = data;

  return [
    <TransactionCard
      key={firstItem[BACKEND_RESPONSE_KEYS.transactionId]}
      accountId={firstItem[BACKEND_RESPONSE_KEYS.accountId]}
      amount={firstItem.amount}
      currentAccountBalance={firstItem.balance}
      showBalance={true}
    />,
    ...rest.map((transaction) => (
      <TransactionCard
        className="mt-4"
        key={transaction[BACKEND_RESPONSE_KEYS.transactionId]}
        accountId={transaction[BACKEND_RESPONSE_KEYS.accountId]}
        amount={transaction.amount}
        currentAccountBalance={transaction.balance}
      />
    )),
  ];
};
