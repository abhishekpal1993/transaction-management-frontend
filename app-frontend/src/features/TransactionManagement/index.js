import { TransactionForm } from "./TransactionForm";
import { TransactionList } from "./TransactionList";
import { SectionHeader } from "../../components/atoms";
import { useTransactionHistory } from "./useTransactionHistory";

export const TransactionManagement = ({
  transactions = [],
  accounts = [],
  isPrefetched = false,
}) => {
  const { transactionList, onSubmitHandler, isLoading } = useTransactionHistory(
    {
      transactions,
      accounts,
      isPrefetched,
    }
  );
  return (
    <div className="grid grid-cols-3 min-h-screen">
      <div className="flex flex-col items-start border py-2 px-4 w-full">
        <SectionHeader className="mb-2">Submit new transaction</SectionHeader>
        <TransactionForm onSubmit={onSubmitHandler} isLoading={isLoading} />
      </div>
      <div className="col-span-2 flex flex-col items-start border py-2 px-4 w-full">
        <SectionHeader className="mb-2">Transaction history</SectionHeader>
        <TransactionList data={transactionList} />
      </div>
    </div>
  );
};
