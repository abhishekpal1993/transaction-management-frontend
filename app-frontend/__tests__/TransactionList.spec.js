import React from "react";
import { render, queries } from "@testing-library/react";
import * as customQueries from "./utils/custom-query";
import { TransactionList } from "../src/features/TransactionManagement/TransactionList";

const renderOptions = {
  queries: { ...queries, ...customQueries },
};

describe("TransactionList > Handling empty transaction lists", () => {
  test("snapshot with empty data", () => {
    const { container } = render(<TransactionList />);
    expect(container).toMatchSnapshot();
  });

  test("data prop is undefined", () => {
    const { getByText } = render(<TransactionList />);
    const notransaction = getByText("No Transactions!");
    expect(notransaction).toBeInTheDocument();
  });

  test("data prop is empty array", () => {
    const { getByText } = render(<TransactionList data={[]} />);
    const notransaction = getByText("No Transactions!");
    expect(notransaction).toBeInTheDocument();
  });

  test("data prop is empty object", () => {
    const { getByText } = render(<TransactionList data={{}} />);
    const notransaction = getByText("No Transactions!");
    expect(notransaction).toBeInTheDocument();
  });
});

describe("TransactionList > Handling non-empty transaction list", () => {
  const sampleDebit = {
    transaction_id: "18e13eed-f05e-4c86-a136-f8ea5ae6ca35",
    account_id: "a5777de7-5604-43d3-8f6c-03acac31e8ca",
    amount: -5,
    balance: 10,
    created_at: "2023-02-21T20:23:06.125272+00:00",
  };
  const sampleCredit = {
    transaction_id: "66dadc89-c8f6-4b2f-9b36-8e81057bc45d",
    account_id: "a5777de7-5604-43d3-8f6c-03acac31e8ca",
    amount: 7,
    balance: 15,
    created_at: "2023-02-21T20:23:03.060149+00:00",
  };

  test("snapshot with sample data", () => {
    const { container } = render(
      <TransactionList data={[sampleDebit, sampleCredit]} />
    );
    expect(container).toMatchSnapshot();
  });

  test("first entry showing correct debit transaction text", () => {
    const { getAllByAccountId } = render(
      <TransactionList data={[sampleDebit]} />,
      renderOptions
    );
    const element = getAllByAccountId(sampleDebit.account_id);
    expect(element[0].querySelector("p:nth-child(1)")).toHaveTextContent(
      `Transferred $${Math.abs(sampleDebit.amount)} from account ${
        sampleDebit.account_id
      }`
    );
  });

  test("first entry showing correct credit transaction text", () => {
    const { getAllByAccountId } = render(
      <TransactionList data={[sampleCredit]} />,
      renderOptions
    );
    const element = getAllByAccountId(sampleCredit.account_id);
    expect(element[0].querySelector("p:nth-child(1)")).toHaveTextContent(
      `Transferred $${Math.abs(sampleCredit.amount)} to account ${
        sampleCredit.account_id
      }`
    );
  });

  test("only first entry have balance text", () => {
    const { getAllByAccountId } = render(
      <TransactionList data={[sampleDebit, sampleCredit]} />,
      renderOptions
    );
    const element = getAllByAccountId(sampleDebit.account_id);
    expect(element[0]).toHaveTextContent(
      `The current account balance $${sampleDebit.balance}`
    );
    expect(element[1]).not.toHaveTextContent(
      `The current account balance $${sampleCredit.balance}`
    );
  });

  test("second entry showing correct credit transaction text", () => {
    const { getAllByAccountId } = render(
      <TransactionList data={[sampleDebit, sampleCredit]} />,
      renderOptions
    );
    const element = getAllByAccountId(sampleCredit.account_id);
    expect(element[1]).not.toHaveTextContent(
      `The current account balance $${sampleCredit.balance}`
    );
  });

  test("second entry showing correct debit transaction text", () => {
    const { getAllByAccountId } = render(
      <TransactionList data={[sampleCredit, sampleDebit]} />,
      renderOptions
    );
    const element = getAllByAccountId(sampleDebit.account_id);
    expect(element[1]).not.toHaveTextContent(
      `The current account balance $${sampleDebit.balance}`
    );
  });
});
