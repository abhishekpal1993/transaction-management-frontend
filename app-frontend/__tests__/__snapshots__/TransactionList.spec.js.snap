// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionList > Handling empty transaction lists snapshot with empty data 1`] = `
<div>
  <div>
    No Transactions!
  </div>
</div>
`;

exports[`TransactionList > Handling non-empty transaction list snapshot with sample data 1`] = `
<div>
  <div
    class="w-full rounded overflow-hidden border border-gray-300 shadow-md px-4 py-2 cursor-default flex flex-col items-start"
    data-account-id="a5777de7-5604-43d3-8f6c-03acac31e8ca"
    data-amount="-5"
    data-balance="10"
    data-type="transaction"
  >
    <p
      class="text-gray-700 text-base"
    >
      Transferred $5 from account a5777de7-5604-43d3-8f6c-03acac31e8ca
    </p>
    <p
      class="text-gray-700 text-base mt-4"
    >
      The current account balance $10
    </p>
  </div>
  <div
    class="w-full rounded overflow-hidden border border-gray-300 shadow-md px-4 py-2 cursor-default flex flex-col items-start mt-4"
    data-account-id="a5777de7-5604-43d3-8f6c-03acac31e8ca"
    data-amount="7"
    data-balance="15"
    data-type="transaction"
  >
    <p
      class="text-gray-700 text-base"
    >
      Transferred $7 to account a5777de7-5604-43d3-8f6c-03acac31e8ca
    </p>
  </div>
</div>
`;
